Solutions/Microsoft 365/Analytic Rules/SharePoint_Downloads_byNewIP.yaml

```diff
@@ -18,35 +17,41 @@ tactics:
   let szSharePointFileOperation = "SharePointFileOperation";
+  // Define an array of SharePoint operations of interest
   let szOperations = dynamic(["FileDownloaded", "FileUploaded"]);
+  // Define the start and end time for the analysis period
   let starttime = 14d;
   let endtime = 1d;
-  let historicalActivity =
-  OfficeActivity
+  // Define a baseline of normal user behavior
+  let userBaseline = OfficeActivity
   | where TimeGenerated between(ago(starttime)..ago(endtime))
   | where RecordType =~ szSharePointFileOperation
   | where Operation in~ (szOperations)
-  | summarize historicalCount = count() by ClientIP, RecordType, Operation;
-  let recentActivity = OfficeActivity
+  | where isnotempty(UserAgent)
+  | extend GeoLocation = geo_info_from_ip_address(ClientIP)
+  | summarize Count = count() by UserId, Operation, tostring(GeoLocation), Site_Url, ClientIP
+  | summarize AvgCount = avg(Count) by UserId, Operation, GeoLocation, Site_Url, ClientIP;
+  // Get recent user activity
+  let recentUserActivity = OfficeActivity
   | where TimeGenerated > ago(endtime)
   | where RecordType =~ szSharePointFileOperation
   | where Operation in~ (szOperations)
-  | summarize min(Start_Time), max(Start_Time), recentCount = count() by ClientIP, RecordType, Operation;
-  let RareIP = recentActivity | join kind= leftanti ( historicalActivity ) on ClientIP, RecordType, Operation
-  // More than 50 downloads/uploads from a new IP
-  | where recentCount > threshold;
-  OfficeActivity
-  | where TimeGenerated >= ago(endtime)
-  | where RecordType =~ szSharePointFileOperation
-  | where Operation in~ (szOperations)
-  | join kind= inner (RareIP) on ClientIP, RecordType, Operation
-  | where Start_Time between(min_Start_Time .. max_Start_Time)
-  | summarize StartTimeUtc = min(min_Start_Time), EndTimeUtc = max(max_Start_Time) by RecordType, Operation, UserType, UserId, ClientIP, OfficeWorkload, Site_Url, OfficeObjectId, UserAgent, IPSeenCount = recentCount
+  | where isnotempty(UserAgent)
+  | extend GeoLocation = geo_info_from_ip_address(ClientIP)
+  | summarize StartTimeUtc = min(TimeGenerated), EndTimeUtc = max(TimeGenerated), RecentCount = count() by UserId, UserType, Operation, tostring(GeoLocation), Site_Url, ClientIP, OfficeObjectId, OfficeWorkload, UserAgent;
+  // Join the baseline and recent activity, and calculate the deviation
+  let UserBehaviorAnalysis = userBaseline | join kind=inner (recentUserActivity) on UserId, Operation, GeoLocation, Site_Url, ClientIP
+  | extend Deviation = abs(RecentCount - AvgCount) / AvgCount;
+  // Filter for significant deviations
+  UserBehaviorAnalysis
+  | where Deviation > threshold
+  | project StartTimeUtc, EndTimeUtc, UserId, UserType, Operation, ClientIP, tostring(GeoLocation), Site_Url, OfficeObjectId, OfficeWorkload, UserAgent, Deviation, Count=RecentCount
   | extend timestamp = StartTimeUtc, AccountCustomEntity = UserId, IPCustomEntity = ClientIP, URLCustomEntity = Site_Url
```

Comment:
We don't need to have these *CustomEntity columns now - I know this wasn't the focus of your work here but whilst you are updating can you please remove these columns and update the entity mappings to just map to the underlying columns.